backends/dpdk/dpdkArch.cpp

```diff
@@ -930,4 +930,129 @@ const IR::Node *PrependPDotToActionArgs::preorder(IR::PathExpression *path) {
     return path;
 }
 
+/* This function updates the control block with newly created declarations and
+   statements for table match keys */
+const IR::Node *copyMatchKeysToSingleStruct::postorder(IR::P4Control *c) {
+    auto control = getOriginal();
+    injector.inject_control(control, c);
+    if (controlBlock && c->name == controlBlock) {
+        for (auto sm : stmt_map) {
+            if (sm.first == c->name) {
+                if (sm.second.empty())
+                    return c;
+                sm.second.append(c->body->components);
+                auto newBody = new IR::BlockStatement(c->body->annotations, sm.second);
+                c->body = newBody;
+                return c;
+            }
+        }
+    }
+    return c;
+}
+
+/* This function transforms the table so that all match keys come from the same struct.
+   Mirror copies of match fields are created in metadata struct and table is updated to
+   use the metadata fields.
+
+   control ingress(inout headers h, inout metadata m) {
+   {
+       ...
+       table tbl {
+           key = {
+               hdr.ethernet.srcAddr : lpm;
+               hdr.ipv4.totalLen : exact;
+           }
+           ...
+       }
+       apply {
+           tbl.apply();
+       }
+   }
+
+   gets translated to
+   control ingress(inout headers h, inout metadata m) {
+       bit<48> tbl_ethernet_srcAddr;  // These declarations are later copied to metadata struct
+       bit<16> tbl_ipv4_totalLen;     // in CollectLocalVariableToMetadata pass.
+       ...
+       table tbl {
+           key = {
+               tbl_ethernet_srcAddr: lpm;
+               tbl_ipv4_totalLen   : exact;
+           }
+	   ...
+       }
+       apply {
+           tbl_ethernet_srcAddr = h.ethernet.srcAddr;
```

Comment:
this kind of change is not trivial, since the table apply call may be part of an expression.
Most such expressions may be normalized by prior midend passes, but the default midends we use tend to leave alone table.apply calls within switch and if statements. The ones in switch statements in particular cannot be moved out.

Reply:
Please note that this problem is tackled by the KeysSideEffect pass. You may want to reuse the code there if you want to insert copy statements in the right places. The code there detects where the insertions must be performed.

Reply:
Please add a test case where a table apply is within an if or switch.

Reply:
If there is not already such a test case, I would also recommend adding a test case where an earlier table has an action that can _modify_ one of the header fields of a packet, that is then later matched as a table key field, and verify that any transformation you do modifies not only the original header field in hdr.ipv4.dstAddress (for example), but also in any copies that your transformation creates.

Reply:
@mbudiu-vmw Thanks for pointing to the KeySideEffect pass. I will update this PR with some code reused from this pass.

Reply:
Will add a few tests as well for covering the above mentioned scenarios.